File: 200. Number of Islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    if ( grid == null || grid.length == 0 || grid[0].length == 0 ) {
        return 0
    }

    var m = grid.length;
    var n = grid[0].length;
    var visited = [];
    var island = 0;

    for ( var i = 0; i < m; i++ ) {
        visited[i] = [];
        for ( var j = 0; j < n; j++ ) {
            visited[i][j] = false;
        }
    }

    for ( var i = 0; i < m; i++ ) {
        for ( var j = 0; j < n; j++ ) {
            if ( !visited[i][j] && grid[i][j] == 1 ){
                island ++;
                bfs( grid, visited, i, j );
            }
        }
    }

    return island;

    function bfs ( grid, visited, i, j ) {
        if ( i >= 0 && i < grid.length
            && j >=0 && j < grid[0].length
            && !visited[i][j]
            && grid[i][j] == 1 ) {
            visited[i][j] = true;
            bfs( grid, visited, i - 1, j );
            bfs( grid, visited, i, j - 1 );
            bfs( grid, visited, i + 1, j );
            bfs( grid, visited, i, j + 1 );
        }
    }
};

File: 200. Number of Islands.py
class Solution_1(object):
    def numIslands(self, grid):
        if not grid or len(grid) == 0 or len(grid[0]) == 0:
            return 0

        m = len(grid)
        n = len(grid[0])

        visited = [[False for i in xrange(n)] for i in xrange(m)]
        result = 0

        for i in xrange(m):
            for j in xrange(n):
                if visited[i][j] == False and grid[i][j] == '1':
                    result += 1
                    self.bfs(grid, visited, i, j)

        return result

    def bfs(self, grid, visited, i, j):
        if i >= 0 and i < len(grid) and j >= 0 and j < len(
                grid[0]) and visited[i][j] == False and grid[i][j] == '1':
            visited[i][j] = True
            self.bfs(grid, visited, i - 1, j)
            self.bfs(grid, visited, i + 1, j)
            self.bfs(grid, visited, i, j - 1)
            self.bfs(grid, visited, i, j + 1)

"""
Solution 2
"""
class Solution2:
    """
    @param grid: a boolean 2D matrix
    @return: an integer
    """
    def numIslands(self, grid):
        if grid is None or len(grid) == 0 or len(grid[0]) == 0:
            return 0
        result = 0
        dx = [1, 0, -1, 0]
        dy = [0, 1, 0, -1]
        for i in xrange(len(grid)):
            for j in xrange(len(grid[0])):
                if grid[i][j] == 1:
                    result += 1
                    q = [(i, j)]
                    while q:
                        x, y = q.pop(0)
                        grid[x][y] = 2
                        for k in xrange(4):
                            newX = x + dx[k]
                            newY = y + dy[k]
                            if newX >= 0 and newX < len(grid) and newY >= 0 and newY < len(grid[0]) and grid[newX][newY] == 1:
                                q.append((newX, newY))
        return result